File: algorithms/sorting/counting.py
"""
=====================================================================
Counting Sort / Сортировка подсчётом
---------------------------------------------------------------------
Идея: считаем частоты значений, восстанавливаем массив.
Time:  O(n + k), k = range
Space: O(k)
Stable: yes
In-place: no
Условие: целые в ограниченном диапазоне.
=====================================================================
"""
from typing import List


def counting_sort(arr: List[int]) -> List[int]:
    if not arr:
        return arr

    lo, hi = min(arr), max(arr)
    rng = hi - lo + 1

    # 1. Частоты
    count = [0] * rng
    for num in arr:
        count[num - lo] += 1

    # 2. Префиксные суммы — для стабильности
    for i in range(1, rng):
        count[i] += count[i - 1]

    # 3. Восстановление с конца (стабильность)
    output = [0] * len(arr)
    for num in reversed(arr):
        count[num - lo] -= 1
        output[count[num - lo]] = num

    return output


if __name__ == "__main__":
    print(counting_sort([4, 2, 2, 8, 3, 3, 1]))  # [1, 2, 2, 3, 3, 4, 8]

File: algorithms/sorting/counting.js
// =====================================================================
// Counting Sort / Сортировка подсчётом
// ---------------------------------------------------------------------
// Идея: считаем количество вхождений каждого значения, потом
//       восстанавливаем отсортированный массив по этой статистике.
// Time:  O(n + k), где k — диапазон значений (max - min + 1)
// Space: O(k)
// Stable: yes (при правильной реализации с префиксными суммами)
// In-place: no
// Условие: целые числа в ограниченном диапазоне. На вещественных не работает.
// =====================================================================

function countingSort(arr) {
    if (arr.length === 0) return arr;

    const min = Math.min(...arr);
    const max = Math.max(...arr);
    const range = max - min + 1;

    // 1. Считаем частоты
    const count = new Array(range).fill(0);
    for (const num of arr) {
        count[num - min]++;
    }

    // 2. Префиксные суммы — это даёт стабильность.
    //    count[i] = индекс конца блока значения (i + min)
    for (let i = 1; i < range; i++) {
        count[i] += count[i - 1];
    }

    // 3. Восстановление. Идём с конца, чтобы сохранить порядок равных
    const output = new Array(arr.length);
    for (let i = arr.length - 1; i >= 0; i--) {
        const num = arr[i];
        count[num - min]--;
        output[count[num - min]] = num;
    }

    return output;
}

console.log(countingSort([4, 2, 2, 8, 3, 3, 1])); // [1, 2, 2, 3, 3, 4, 8]

module.exports = countingSort;
